android/expoview/src/main/java/host/exp/exponent/RNObject.kt

```diff
@@ -0,0 +1,314 @@
+// Copyright 2015-present 650 Industries. All rights reserved.
+package host.exp.exponent
+
+import host.exp.exponent.analytics.EXL
+import host.exp.expoview.BuildConfig
+import java.lang.reflect.Constructor
+import java.lang.reflect.Field
+import java.lang.reflect.InvocationTargetException
+import java.lang.reflect.Method
+
+// TODO: add type checking in DEBUG
+class RNObject {
+  private val className: String // Unversioned
+  private var clazz: Class<*>? = null // Versioned
+  private var instance: Any? = null // Versioned
+
+  // We ignore the version of clazz
+  constructor(clazz: Class<*>?) {
+    className = removeVersionFromClass(clazz)
+  }
+
+  constructor(className: String) {
+    this.className = className
+  }
+
+  private constructor(obj: Any?) {
+    assign(obj)
+    className = removeVersionFromClass(clazz)
+  }
+
+  val isNull: Boolean
+    get() = instance == null
+  val isNotNull: Boolean
+    get() = instance != null
+
+  // required for "unversioned" flavor check
+  fun loadVersion(version: String): RNObject {
+    try {
+      clazz = if (version == UNVERSIONED || BuildConfig.FLAVOR == "unversioned") {
+        if (className.startsWith("host.exp.exponent")) {
+          Class.forName("versioned.$className")
+        } else {
+          Class.forName(className)
+        }
+      } else {
+        Class.forName("abi" + version.replace('.', '_') + '.' + className)
+      }
+    } catch (e: ClassNotFoundException) {
+      EXL.e(TAG, e)
+    }
+    return this
+  }
+
+  fun assign(obj: Any?) {
+    if (obj != null) {
+      clazz = obj.javaClass
+    }
+    instance = obj
+  }
+
+  fun get(): Any? {
+    return instance
+  }
+
+  fun rnClass(): Class<*>? {
+    return clazz
+  }
+
+  fun version(): String {
+    val versionedClassName = clazz!!.name
+    return if (versionedClassName.startsWith("abi")) {
+      val abiVersion = versionedClassName.split(".").toTypedArray()[0]
+      abiVersion.substring(3)
+    } else {
+      UNVERSIONED
+    }
+  }
+
+  fun construct(vararg args: Any?): RNObject {
+    try {
+      instance = getConstructorWithTypes(clazz, *objectsToClasses(*args)).newInstance(*args)
+    } catch (e: NoSuchMethodException) {
+      EXL.e(TAG, e)
+    } catch (e: InvocationTargetException) {
+      EXL.e(TAG, e)
+    } catch (e: InstantiationException) {
+      EXL.e(TAG, e)
+    } catch (e: IllegalAccessException) {
+      EXL.e(TAG, e)
+    }
+    return this
+  }
+
+  fun call(name: String, vararg args: Any?): Any? {
+    return callWithReceiver(instance, name, *args)
+  }
+
+  fun callRecursive(name: String, vararg args: Any?): RNObject? {
+    val result = call(name, *args) ?: return null
+    return wrap(result)
```

Comment:
I think it is better to make clear this is a static method.
```suggestion
    return RNObject.wrap(result)
```

Here it doesn't matter so much, but it is especially confusing with:

```
val thread = new Thread()
thread.sleep() // does not sleep "thread", calls Thread.sleep()
```

Reply:
Unfortunately intellij complains that it's a redundant qualifier since it's called from within. Going to opt to omit it even though I agree it's slightly less clear.